CMakeLists.txt

```diff
@@ -26,82 +24,162 @@ if(HIOP_USE_MPI)
+                                        # duplicated code
+            -Wuseless-cast              # Warn when casting to the same type
+    )
+endif()
+
+target_link_libraries(hiop_math INTERFACE hiop_options hiop_warnings)
+
 if(HIOP_USE_GPU)
+  include(CheckLanguage)
+  enable_language(CUDA)
+  check_language(CUDA)
+  if(NOT DEFINED CMAKE_CUDA_STANDARD)
+    set(CMAKE_CUDA_STANDARD 11)
+    set(CMAKE_CUDA_STANDARD_REQUIRED ON)
+  endif()
+
+  if(HIOP_NVCC_ARCH)
+      set(CMAKE_CUDA_FLAGS "-arch=${HIOP_NVCC_ARCH} --expt-extended-lambda")
+      message(STATUS "Using CUDA arch ${HIOP_NVCC_ARCH}")
+  else()
+      set(CMAKE_CUDA_FLAGS "-arch=sm_35 --expt-extended-lambda")
+      message(STATUS "Using CUDA arch sm_35")
+  endif()
+
   # for now we rely on MAGMA for GPUs computations
-  set(MAGMA_LIBRARIES "-lmagma")
+  target_link_libraries(hiop_magma INTERFACE "-lmagma")
   if(HIOP_MAGMA_DIR)
-    set(MAGMA_LIBRARIES "-L${HIOP_MAGMA_DIR}/lib ${MAGMA_LIBRARIES}")
-    set(MAGMA_INCLUDE "${HIOP_MAGMA_DIR}/include")
+    target_link_libraries(hiop_magma INTERFACE "-L${HIOP_MAGMA_DIR}/lib")
+    target_include_directories(hiop_magma INTERFACE "${HIOP_MAGMA_DIR}/include")
   endif(HIOP_MAGMA_DIR)
 
-  set(HIOP_CUDA_LIBRARIES "-lculibos -lcublas -lcublasLt -lnvblas -lcusparse -lcudart -lcudadevrt")
-  find_package(CUDA 10.1)
-  
   if(HIOP_CUDA_LIB_DIR)
-    set(HIOP_CUDA_LIBRARIES "-L${HIOP_CUDA_LIB_DIR} ${HIOP_CUDA_LIBRARIES}")
+    target_link_libraries(hiop_cuda INTERFACE "-L${HIOP_CUDA_LIB_DIR}")
   else()
-    set(HIOP_CUDA_LIBRARIES "-L${CUDA_TOOLKIT_ROOT_DIR}/lib64 ${HIOP_CUDA_LIBRARIES}")
+    target_link_libraries(hiop_cuda INTERFACE "-L${CUDA_TOOLKIT_ROOT_DIR}")
   endif(HIOP_CUDA_LIB_DIR)
 
   # on some systems RHEL Nvidia toolkit puts cuBlas in different locations
   # user can setup/overwrite the lib directory
   if(HIOP_CUBLAS_LIB_DIR)
-    set(HIOP_CUDA_LIBRARIES "-L${HIOP_CUBLAS_LIB_DIR} ${HIOP_CUDA_LIBRARIES}")
+    target_link_libraries(hiop_cuda INTERFACE "-L${HIOP_CUBLAS_LIB_DIR}")
   endif(HIOP_CUBLAS_LIB_DIR)
 
   if(HIOP_CUDA_INCLUDE_DIR)
-    set(HIOP_CUDA_INCLUDE "${HIOP_CUDA_INCLUDE_DIR}")
+    target_include_directories(hiop_cuda INTERFACE "${HIOP_CUDA_INCLUDE_DIR}")
   else()
-    set(HIOP_CUDA_INCLUDE "${CUDA_INCLUDE_DIRS}")
+    target_include_directories(hiop_cuda INTERFACE "${CUDA_INCLUDE_DIR}")
   endif(HIOP_CUDA_INCLUDE_DIR)
 
+  target_link_libraries(hiop_math INTERFACE hiop_magma hiop_cuda)
   message("HiOp support for GPUs is on")
-  message("HiOp GPU libs: '${MAGMA_LIBRARIES} ${HIOP_CUDA_LIBRARIES}'") 
-  message("HiOp GPU includes: '${MAGMA_INCLUDE} ${HIOP_CUDA_INCLUDE}'") 
-  include_directories(${MAGMA_INCLUDE} ${HIOP_CUDA_INCLUDE})
-  set(HIOP_MATH_LIBRARIES ${MAGMA_LIBRARIES} ${HIOP_CUDA_LIBRARIES} ${HIOP_MATH_LIBRARIES})
+  get_target_property(cuda_includes hiop_cuda INTERFACE_INCLUDE_DIRECTORIES)
+  get_target_property(magma_includes hiop_magma INTERFACE_INCLUDE_DIRECTORIES)
```

Comment:
something wrong with magma_includes and cuda_includes on my redhat desktop. 

when I make, the include directories of magma and cuda do not appear under -I and the compilation fails

Reply:
a) Do the libraries show up under the respective messages during cmake configuration?
b) Are you specifying them manually with `HIOP_MAGMA_DIR`, `HIOP_CUDA_LIB_DIR`, etc?

Reply:
setting `HIOP_MAGMA_DIR` or `HIOP_CUDA_LIB_DIR` manually or not does not make a difference.

after cmake invoke they appear ok
`
CUDA include directories: /home/petra1/work/projects/hiop/ /home/petra1/work/installs/magma-2.5.2/include
CUDA linked libraries: -L -lmagma;-L/home/petra1/work/installs/magma-2.5.2/lib
`

after make
`
In file included from /home/petra1/work/projects/hiop/src/Optimization/hiopKKTLinSysDense.hpp:53:0,
                 from /home/petra1/work/projects/hiop/src/Optimization/hiopAlgFilterIPM.cpp:51:
/home/petra1/work/projects/hiop/src/LinAlg/hiopLinSolver.hpp:230:22: fatal error: magma_v2.h: No such file or directory
 #include "magma_v2.h"
`

Reply:
@cnpetra what is the CMake version you use?

Reply:
cmake 3.14.7

Reply:
@cnpetra, I am having difficulties reproducing the problem you see, as well. I downgraded to CMake 3.14 just in case, but everything still works. I tested installation on Power9/V100. Modules loaded are:
```shell
$ module list
Currently Loaded Modulefiles:
  1) gcc/7.4.0              4) ipopt/3.13.0           7) metis/5.1.0
  2) cuda/10.2              5) magma/2.5.2_cuda10.2
  3) openmpi/3.1.5          6) cmake/3.14.3
```
I used my own build of UMFPACK. I configured build with:
```shell
CC=mpicc CXX=mpicxx FC=mpif90 cmake ../hiop/ \
-DHIOP_USE_GPU=On \
-DHIOP_WITH_KRON_REDUCTION=On \
-DUMFPACK_DIR=/path/to/suitesparse \
-DMETIS_DIR=/path/to/metis/5.1.0/ \
-DHIOP_MAGMA_DIR=/path/to/magma/2.5.2/cuda10.2/
```
After this, the code builds cleanly. All tests pass:
```shell
$ make test
Running tests...
Test project /people/pele118/src/hiop_asher/build_newell
      Start  1: NlpDenseCons1_5H
 1/12 Test  #1: NlpDenseCons1_5H .................   Passed    0.45 sec
      Start  2: NlpDenseCons1_5K
 2/12 Test  #2: NlpDenseCons1_5K .................   Passed    0.48 sec
      Start  3: NlpDenseCons1_50K
 3/12 Test  #3: NlpDenseCons1_50K ................   Passed    1.24 sec
      Start  4: NlpDenseCons1_50K_mpi
 4/12 Test  #4: NlpDenseCons1_50K_mpi ............   Passed    0.84 sec
      Start  5: NlpDenseCons2_5H
 5/12 Test  #5: NlpDenseCons2_5H .................   Passed    0.43 sec
      Start  6: NlpDenseCons2_5K
 6/12 Test  #6: NlpDenseCons2_5K .................   Passed    0.81 sec
      Start  7: NlpDenseCons3_5H
 7/12 Test  #7: NlpDenseCons3_5H .................   Passed    0.48 sec
      Start  8: NlpDenseCons3_5K
 8/12 Test  #8: NlpDenseCons3_5K .................   Passed    1.01 sec
      Start  9: NlpDenseCons3_50K
 9/12 Test  #9: NlpDenseCons3_50K ................   Passed    8.10 sec
      Start 10: NlpDenseCons3_50K_mpi
10/12 Test #10: NlpDenseCons3_50K_mpi ............   Passed    4.37 sec
      Start 11: NlpMixedDenseSparse_1
11/12 Test #11: NlpMixedDenseSparse_1 ............   Passed    0.60 sec
      Start 12: NlpMixedDenseSparse_2
12/12 Test #12: NlpMixedDenseSparse_2 ............   Passed    0.60 sec

100% tests passed, 0 tests failed out of 12

Total Test time (real) =  19.54 sec
```
Can you please provide more details about your system and how you configure CMake?

Reply:
```
rm -rf *
cmake3 -DHIOP_USE_GPU=ON -DHIOP_MAGMA_DIR=/home/petra1/work/installs/magma-2.5.2/ -DHIOP_CUDA_INCLUDE_DIR=/usr/local/cuda-10.2/include/ -DHIOP_CUDA_LIB_DIR=/usr/local/cuda-10.2/lib64 -DHIOP_CUBLAS_LIB_DIR=/usr/local/cuda-10.2/targets/x86_64-linux/lib/lib64 ..
```

Reply:
to help you debug, if I add the includes above manually, that is
```
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -I/usr/local/cuda-10.2/include/ -I/home/petra1/work/installs/magma-2.5.2//include")
```
it compiles perfectly

Reply:
Try removing `" "` around `"${HIOP_MAGMA_DIR}/include"`.

Reply:
it does not fix. I will approve and merge this PR though. My RH box is a bit outdated and I don't think it worth the effort of chasing since it works on all other platforms... @pelesh let me know when/if you approve of this.